pkg/promtail/client/client.go

```diff
@@ -111,21 +109,29 @@ func newMetrics(reg prometheus.Registerer) *metrics {
 	}
 
 	if reg != nil {
-		reg.MustRegister(
-			m.encodedBytes,
-			m.sentBytes,
-			m.droppedBytes,
-			m.sentEntries,
-			m.droppedEntries,
-			m.requestDuration,
-			m.batchRetries,
-			m.streamLag,
-		)
+		m.encodedBytes = mustRegisterOrGet(reg, m.encodedBytes).(*prometheus.CounterVec)
+		m.sentBytes = mustRegisterOrGet(reg, m.sentBytes).(*prometheus.CounterVec)
+		m.droppedBytes = mustRegisterOrGet(reg, m.droppedBytes).(*prometheus.CounterVec)
+		m.sentEntries = mustRegisterOrGet(reg, m.sentEntries).(*prometheus.CounterVec)
+		m.droppedEntries = mustRegisterOrGet(reg, m.droppedEntries).(*prometheus.CounterVec)
+		m.requestDuration = mustRegisterOrGet(reg, m.requestDuration).(*prometheus.HistogramVec)
+		m.batchRetries = mustRegisterOrGet(reg, m.batchRetries).(*prometheus.CounterVec)
+		m.streamLag = mustRegisterOrGet(reg, m.streamLag).(*metric.Gauges)
 	}
 
 	return &m
 }
 
+func mustRegisterOrGet(reg prometheus.Registerer, c prometheus.Collector) prometheus.Collector {
```

Comment:
Would it be good to not return the `prometheus.Collector`? It would simplify things and make it clear that the method is just for registering the collector without modifying it.

Reply:
I need it actually so all client can increment the metrics.